# RQ4


In this RQ, We want to find ten players with the highest ratio between completed passes and toatal passes

In [2]:
import json
import pandas as pd

In [3]:
with open(r"C:\Users\Nino\Desktop\events_England.json") as myfile :
    data = myfile.read()
events = pd.DataFrame(json.loads(data))


In [5]:
events.head(5)

eventId eventName   eventSec         id  matchId matchPeriod  playerId  \
0        8      Pass   2.758649  177959171  2499719          1H     25413   
1        8      Pass   4.946850  177959172  2499719          1H    370224   
2        8      Pass   6.542188  177959173  2499719          1H      3319   
3        8      Pass   8.143395  177959174  2499719          1H    120339   
4        8      Pass  10.302366  177959175  2499719          1H    167145   

                                  positions subEventId subEventName  \
0  [{'y': 49, 'x': 49}, {'y': 78, 'x': 31}]         85  Simple pass   
1  [{'y': 78, 'x': 31}, {'y': 75, 'x': 51}]         83    High pass   
2  [{'y': 75, 'x': 51}, {'y': 71, 'x': 35}]         82    Head pass   
3  [{'y': 71, 'x': 35}, {'y': 95, 'x': 41}]         82    Head pass   
4  [{'y': 95, 'x': 41}, {'y': 88, 'x': 72}]         85  Simple pass   

             tags  teamId  
0  [{'id': 1801}]    1609  
1  [{'id': 1801}]    1609  
2  [{'id': 1801}]    1609  
3  [{'id': 1801}]    1609  
4  [{'id': 1801}]    1609

since in this RQ we are just interested in "pass" events, we filter the rows that has eventId=8

In [6]:
pass_events = events[events.eventId == 8]
pass_events

eventId eventName     eventSec         id  matchId matchPeriod  \
0             8      Pass     2.758649  177959171  2499719          1H   
1             8      Pass     4.946850  177959172  2499719          1H   
2             8      Pass     6.542188  177959173  2499719          1H   
3             8      Pass     8.143395  177959174  2499719          1H   
4             8      Pass    10.302366  177959175  2499719          1H   
5             8      Pass    12.548934  177959177  2499719          1H   
6             8      Pass    13.961228  177959186  2499719          1H   
9             8      Pass    15.320341  177959178  2499719          1H   
10            8      Pass    18.051875  177959179  2499719          1H   
11            8      Pass    20.426527  177959180  2499719          1H   
14            8      Pass    24.588310  177959182  2499719          1H   
15            8      Pass    25.904249  177959194  2499719          1H   
16            8      Pass    27.527696  177959195  2499719          1H   
17            8      Pass    29.980898  177959196  2499719          1H   
18            8      Pass    31.163870  177959197  2499719          1H   
19            8      Pass    32.805618  177959202  2499719          1H   
20            8      Pass    35.685082  177959184  2499719          1H   
21            8      Pass    36.661264  177959204  2499719          1H   
24            8      Pass    39.940109  177959206  2499719          1H   
25            8      Pass    43.045831  177959185  2499719          1H   
26            8      Pass    44.425679  177959187  2499719          1H   
37            8      Pass    76.625696  177959199  2499719          1H   
42            8      Pass    86.773457  177959203  2499719          1H   
43            8      Pass    89.249387  177959207  2499719          1H   
44            8      Pass    93.115548  177959208  2499719          1H   
45            8      Pass    93.961235  177959211  2499719          1H   
50            8      Pass   148.972747  177959220  2499719          1H   
65            8      Pass   198.850021  177959228  2499719          1H   
70            8      Pass   215.982359  177959232  2499719          1H   
71            8      Pass   217.803357  177959234  2499719          1H   
...         ...       ...          ...        ...      ...         ...   
643093        8      Pass  2638.826239  251596386  2500098          2H   
643096        8      Pass  2642.981466  251596388  2500098          2H   
643097        8      Pass  2646.816333  251596134  2500098          2H   
643102        8      Pass  2668.905130  251596392  2500098          2H   
643103        8      Pass  2676.466810  251596393  2500098          2H   
643104        8      Pass  2679.856529  251596394  2500098          2H   
643105        8      Pass  2682.271344  251596395  2500098          2H   
643106        8      Pass  2683.754085  251596396  2500098          2H   
643107        8      Pass  2684.747496  251596397  2500098          2H   
643110        8      Pass  2703.287981  251596178  2500098          2H   
643118        8      Pass  2714.004168  251596188  2500098          2H   
643119        8      Pass  2726.496538  251596193  2500098          2H   
643120        8      Pass  2730.025560  251596196  2500098          2H   
643121        8      Pass  2732.694515  251596198  2500098          2H   
643122        8      Pass  2733.173494  251596199  2500098          2H   
643124        8      Pass  2741.371988  251596204  2500098          2H   
643125        8      Pass  2742.418880  251596206  2500098          2H   
643126        8      Pass  2743.534946  251596207  2500098          2H   
643127        8      Pass  2746.848378  251596209  2500098          2H   
643128        8      Pass  2748.321677  251596211  2500098          2H   
643129        8      Pass  2750.561883  251596212  2500098          2H   
643130        8      Pass  2753.232101  251596214  2500098          2H   
643131      

Now we want to know about accurate passes so we filter rows that contain tag id 1801

In [7]:
acc_passes = pass_events[pass_events['tags'].apply(lambda x: pd.Series([dict['id'] for dict in x]).eq(1801).any())]

In [9]:
acc_passes.head(5)

eventId eventName   eventSec         id  matchId matchPeriod  playerId  \
0        8      Pass   2.758649  177959171  2499719          1H     25413   
1        8      Pass   4.946850  177959172  2499719          1H    370224   
2        8      Pass   6.542188  177959173  2499719          1H      3319   
3        8      Pass   8.143395  177959174  2499719          1H    120339   
4        8      Pass  10.302366  177959175  2499719          1H    167145   

                                  positions subEventId subEventName  \
0  [{'y': 49, 'x': 49}, {'y': 78, 'x': 31}]         85  Simple pass   
1  [{'y': 78, 'x': 31}, {'y': 75, 'x': 51}]         83    High pass   
2  [{'y': 75, 'x': 51}, {'y': 71, 'x': 35}]         82    Head pass   
3  [{'y': 71, 'x': 35}, {'y': 95, 'x': 41}]         82    Head pass   
4  [{'y': 95, 'x': 41}, {'y': 88, 'x': 72}]         85  Simple pass   

             tags  teamId  
0  [{'id': 1801}]    1609  
1  [{'id': 1801}]    1609  
2  [{'id': 1801}]    1609  
3  [{'id': 1801}]    1609  
4  [{'id': 1801}]    1609

Now we calculate the ratio of accurate passes

In [11]:
ratio = ( acc_passes["playerId"].value_counts() / pass_events["playerId"].value_counts() ).sort_values(ascending=[False])
ratio

532949    1.000000
423162    1.000000
25703     1.000000
26109     1.000000
134708    1.000000
227756    1.000000
447538    1.000000
291528    1.000000
447254    1.000000
340476    1.000000
353241    1.000000
415747    1.000000
9380      0.975914
14886     0.956522
8324      0.948020
8464      0.945076
8307      0.943206
135103    0.940217
7856      0.939914
7918      0.938505
70086     0.937584
212651    0.937500
120339    0.936451
14808     0.935556
71703     0.935528
11152     0.932601
71654     0.932449
8745      0.931818
74        0.930973
241564    0.930931
            ...   
9123      0.677262
9414      0.675302
293686    0.673203
8903      0.670000
145692    0.669811
8416      0.667297
20606     0.666667
15301     0.666667
12829     0.665037
224206    0.662791
8013      0.658537
49971     0.657321
8686      0.651685
8677      0.650510
7961      0.649402
531655    0.636364
245813    0.633136
15054     0.632780
55224     0.623529
337225    0.621622
32968     0.611940
280157    0.

As you can see some of our rows have value 1. lets evaluate one of this points :

In [16]:
print(acc_passes["playerId" ].value_counts()[26109])
print(pass_events["playerId" ].value_counts()[26109])

2
2


This player had only 2 passes and all of them were accurate so we can't judge whether he is a good palyer or not based on this information. Now we put a threshhold to solve this problem. for setting a threshold, it's good to have an initial understanding of our data:

In [17]:
pass_events["playerId"].value_counts().describe()

count     514.000000
mean      639.410506
std       565.180743
min         1.000000
25%       157.000000
50%       531.000000
75%       968.500000
max      2974.000000
Name: playerId, dtype: float64

As you can see the first quartile is 157 and it means that 75% of our population had more than 157 passes, also the mean is 639, so it seems fair to set the threshold equal to 157. let's see :

In [19]:
countA = pass_events.playerId.value_counts()
countB = countA[countA.apply(lambda x : x > 157)]
countB.head(10)

49876     2974
70086     2964
105339    2842
25553     2713
38021     2672
70122     2456
8317      2382
48        2370
8277      2316
54        2196
Name: playerId, dtype: int64

Now, I calculate the ratio :

In [25]:
rate = (acc_passes["playerId"].value_counts()/ countB).sort_values(ascending=[False]).dropna()
rate[0:10]

9380      0.975914
8324      0.948020
8464      0.945076
8307      0.943206
135103    0.940217
7856      0.939914
7918      0.938505
70086     0.937584
120339    0.936451
14808     0.935556
Name: playerId, dtype: float64

In [42]:
ratio = pd.DataFrame({"wyId": rate.index , "ratio" : rate.values})
ratio.head(10)

wyId     ratio
0    9380  0.975914
1    8324  0.948020
2    8464  0.945076
3    8307  0.943206
4  135103  0.940217
5    7856  0.939914
6    7918  0.938505
7   70086  0.937584
8  120339  0.936451
9   14808  0.935556

We can also extract the name of these players, so I neet to load players database

In [33]:
with open(r"C:\Users\Nino\Desktop\players.json") as myfile :
    data = myfile.read()
players = pd.DataFrame(json.loads(data))


In [44]:
ratio ["Last name"] = ratio.merge(players, on = "wyId")["lastName"]

In [45]:
ratio.head(10)

wyId     ratio           Last name
0    9380  0.975914              Stones
1    8324  0.948020          Tour\u00e9
2    8464  0.945076               Delph
3    8307  0.943206             Kompany
4  135103  0.940217         Christensen
5    7856  0.939914         Mertesacker
6    7918  0.938505               Jones
7   70086  0.937584            Otamendi
8  120339  0.936451      Elsayed Elneny
9   14808  0.935556  G\u00fcndo\u011fan